README
Software for Locally-Weighted Regression       18 August 1992

William S. Cleveland
Eric Grosse
Ming-Jen Shyu

Locally-weighted regression, or loess, is a procedure for estimating a
regression surface by a multivariate smoothing procedure: fitting a
linear or quadratic function of the independent variables in a moving
fashion that is analogous to how a moving average is computed for a
time series. Compared to classical approaches  - fitting global
parametric functions - loess substantially increases the domain of
surfaces that can be estimated without distortion. Also, a pleasant
fact about loess is that analogues of the statistical procedures used
in parametric function fitting - for example, ANOVA and t intervals -
involve statistics whose distributions are well approximated by
familiar distributions.

The follwing files are included in this distribution.
	README		the instruction file you are reading now
	S.h		header file
	air.c		C source for air data example
	changes		history of changes to loess
	depend.ps	PostScript figure of how routines are related
	ethanol.c	C source for ethanol data example
	galaxy.c	C source for galaxy data example
	gas.c		C source for gas data example
	loess.c		C source (high-level loess routines)
	loess.h		header file for loess_struct and predict_struct
	loess.m		manual page for user-callable loess routines
	loessc.c	C source (low-level loess routines)
	loessf.f	FORTRAN source (low-level loess & predict routines)
	loessf.m	documentation for FORTRAN source
	madeup.c	C source for madeup data example
	makefile	makefile to compile the example codes
	misc.c		C source (anova, pointwise, and other support routines)
	predict.c	C source (high-level predict routines)
	predict.m	manual page for user-callable predict routines
	struct.m	manual page for loess_struct, pred_struct
	supp.f		supplemental Fortran loess drivers

After unpacking these files, just type "make" and if all goes well
you should see output like:

	loess(&gas):
	Number of Observations: 22
	Equivalent Number of Parameters: 5.5
	Residual Standard Error: 0.3404
	
	loess(&gas_null):
	Number of Observations: 22
	Equivalent Number of Parameters: 3.5
	Residual Standard Error: 0.5197
	
	predict(gas_fit_E, m, &gas, &gas_pred):
	1.19641 5.06875 0.523682
	
	pointwise(&gas_pred, m, coverage, &gas_ci):
	1.98562 4.10981 5.48023 5.56651 3.52761 1.71062 1.47205
	1.19641 3.6795 5.05571 5.13526 3.14366 1.19693 0.523682
	0.407208 3.24919 4.63119 4.70401 2.7597 0.683247 -0.424684
	
	anova(&gas_null, &gas, &gas_anova):
	2.5531 15.663 10.1397 0.000860102

To run other examples, simply type "make galaxy", or "make ethanol", etc.

If your loader complains about "-llinpack -lblas" in the makefile, change
it to whatever your system prefers for accessing Linpack and the Blas.
If necessary, these Fortran subroutines can be obtained by
	mail netlib@netlib.bell-labs.com
	send dnrm2 dsvdc dqrdc ddot dqrsl idamax from linpack core.

A 50 page user guide, in PostScript form, is available by anonymous ftp.
        ftp netlib.bell-labs.com
        login: anonymous
        password: <your email address>
        binary
        cd /netlib/a
        get cloess.ps.Z
        quit
        uncompress cloess.ps
This guide describes crucial steps in the proper analysis of data using
loess.  Please read it.

Bug reports are appreciated.  Send electronic mail to
	ehg@netlib.bell-labs.com
including the words "this is not spam" in the Subject line
or send paper mail to
	Eric Grosse
	Bell Labs 2T-502
	Murray Hill NJ 07974
for problems with the Fortran inner core of the algorithm.  
The C drivers were written by Ming-Jen Shyu, who left Bell Labs.  Eric will
fix problems with them when he can.

Remember that this is experimental software distributed free of charge
and comes with no warranty!  Exercise professional caution.

Happy Smoothing!

/*
 * The authors of this software are Cleveland, Grosse, and Shyu.
 * Copyright (c) 1989, 1992 by AT&T.
 * Permission to use, copy, modify, and distribute this software for any
 * purpose without fee is hereby granted, provided that this entire notice
 * is included in all copies of any software which is or includes a copy
 * or modification of this software and in all copies of the supporting
 * documentation for such software.
 * THIS SOFTWARE IS BEING PROVIDED "AS IS", WITHOUT ANY EXPRESS OR IMPLIED
 * WARRANTY.  IN PARTICULAR, NEITHER THE AUTHORS NOR AT&T MAKE ANY
 * REPRESENTATION OR WARRANTY OF ANY KIND CONCERNING THE MERCHANTABILITY
 * OF THIS SOFTWARE OR ITS FITNESS FOR ANY PARTICULAR PURPOSE.
 */



changes
CHANGES PLANNED SOMEDAY
1) more vertices in k-d tree for dimension > 2, to get continuity.
2) triangulation based method.
----------------------

19 Nov 1987 	workspace not big enough for degree=2

22 Jan 1988	switched from depth first to breadth first tree build

14 Mar 1988	lostt.3   extra space needed if (method mod 1000 = 0),
		not the documented (method/1000=0)

28 Apr 1988	l2tr.g 	vval2 needed to be initialized to 0

		galaxy smooth needs double precision on vax

26 May 1988	bbox.g    add 10% margin to allow limited extrapolation

6 June 1988	loess/lostt.f	trL wasn't set if method/1000==0

10 June 1988	losave, loread

		v(RCOND)  1 / max condition number

12 June 1988	lofort

21 June 1988	additional workspace for explicit L

27 June 1988	workspace checking in lowesf was slightly pessimistic

30 June 1988	Changed default fdiam to 0.
		Added warning messages for memory limits and pseudoinverse.

4 Aug 1988	bbox.g  changed margin from 10% to 0.5%.

24 Aug 1988	loser documentation should have specified workspace
		of size ...+m*n, not ...+m**2.

Sep 1988
	loess-based approximations of delta1,2.
	pseudo-values, so statistics are available with robustness iterations.
	reorganize error messages to better fit into S.
	sample driver program.
	somewhat shorter code generated by ehg170.

20 Dec 1988
	workspace in loser

27 Jan 1989
	workspace checking in lostt was a bit pessimistic.

3 Feb 1989
	l2fit, l2tr: error message should contain sqrt(rho)

18 Dec 1989
	ehg141, ehg179-ehg181:   new delta approximations

24 Jan 1990
	master copy moved from Sun3/180 to SGI 4D/240S
	(no intentional changes)

25 Jan 1990
	(many routines touched; ehg127 added)   cleaned up computational
	kernel, added provision for only first dd<=d variables to enter
	the distance calculation ("conditionally parametric variables"),
	added independent bounds on total and componentwise degree for
	local polynomial model,  made extrapolation warning message print
	a bit more detail.

14 Mar 1990
	added setLf argument to lowesd; added lowesr, lowesl for resmoothing.

-------------------------------------------------------
Converting to the new version of loess
5 April 1990

Over the past few months, a number of changes have been made to the
loess package, to provide more control over the local model, to allow
conditionally parametric variables, and to return exact statistical
quantities for the blending method.  Unlike earlier internal
algorithmic improvements, this round of changes added some extra
arguments in the Fortran calling sequences.  The purpose of this note
is to assist in converting programs that called the old version.

An explicit argument setLf has been added to lowesd(), since it affects
the partitioning of the workspace.  To help protect against inadvertent
version mismatches, the version number that lowesd() checks has also
been changed.  The componentwise degree and the specification of
conditionally nonparametric variables can be changed from the default
by modifying iv(CDEG) and iv(NDIST).

The influence matrix L for blending is now explicitly available by
calling a new subroutine lowesl(),  but this loses the speed
advantage of blending.  A faster, sometimes equivalent method is
to use the influence matrix that carries data values to coefficients
at the vertices of the k-d tree.  This information is saved in iv(iv(Lq))
and v(iv(Lf)), for the afficionado.

The new subroutine lowesr() takes advantage of Lq and Lf to allow rapid
resmoothing for applications when only y, not x, is subject to change.
-------------------------------------------------------

7 May 1990
	new delta approximations.
	added prior weights to input format for sample driver.

29 May 1990
	loess,lostt,loser,pseudo moved from Fortran to S.

11 Jul 1990
	column equilibration, so pseudoinverse is needed less often.

27 May 1991
lowesd	version 105;  increased nvmax,ncmax to max(200,n).
l2fit	added ihat=1 (diagL only).
ehg133,lowese	removed unused arguments dist,eta.
ehg190,ehg141	changed name to lowesa, slight change to calling sequence.
ehg144	changed name to lowesc
m9rwt	changed name to lowesw
pseudo	changed name to lowesp

22 Jul 1991    IMPORTANT BUG FIX!
ehg131	vval2 should be dimensioned 0:d, not 0:8

26 Jul 1991
lowesd	change calling sequence to provide tighter memory allocation
diff old/man/internal new/man/internal
< lowesd(105,iv,liv,lv,v,d,n,f,tdeg,setLf)	setup workspace
> lowesd(106,iv,liv,lv,v,d,n,f,tdeg,nvmax,setLf)	setup workspace
< liv	50+(2^d+6)*max(200,n)
< 	if setLf, add nf*max(200,n)
< lv	50+(3*d+4)*max(200,n)+(tau+2)*nf
< 	if setLf, add (d+1)*nf*max(200,n)
> liv	50+(2^d+6)*nvmax
> 	if setLf, add nf*nvmax
> lv	50+(3*d+4)*nvmax+(tau+2)*nf
> 	if setLf, add (d+1)*nf*nvmax
> nvmax	limit on number of vertices for kd-tree; e.g. max(200,n)

20 Sep 1991
sample.f   brought in sync with recent loess changes.

24 Dec 1991
l2fit.f    fixed comment in single precision version

10 Jan 1992
ehg197.f   new formula for approximating trL, valid for small f

15 May 1992
netlib/a/dloess   now includes C drivers (written by Ming-Jen Shyu,
	adapted from code used inside the S system)

22 Jun 1992
ehg191.f     Loop 11 ran too far, picking up one more value than necessary.
	The value was not used, so the loess computation itself is unaffected,
	but on some systems the old code could conceivably cause a reference
	to an invalid memory address and abort with a segmentation fault
	message.

23 Jun 1992
S.h	#include <math.h>,  since loessc.c calls floor() and pow().

18 Aug 1992
netlib/a/dloess   A new release with bug fixes in all the C drivers, new
	example codes, and detail documentations.

25 Mar 1996
predict.c  fix enormous memory leak.       update email address


loessf.m
***************************************************************
* LOESS   smoothing scattered data in one or more variables   *
*         documentation of Fortran routines                   *
*         Cleveland, Devlin, Grosse, Shyu                     *
***************************************************************

1. The typical program would call lowesd, set tolerances in iv,v if
   desired, then call lowesb and lowese.
2. To save the k-d tree, call lowesd, lowesb and then losave; subsequent
   programs would call lohead, set liv and lv, then call loread and lowese.
3. For statistics, get diagL and then call lowesa or get the full hat
   matrix and call lowesc.  Robustness iterations can take advantage of
   lowesw and lowesp.

lowesd(106,iv,liv,lv,v,d,n,f,tdeg,nvmax,setLf)	setup workspace
lowesf(x,y,w,iv,liv,lv,v,m,z,L,hat,s)		slow smooth at z
lowesb(x,y,w,diagL,infl,iv,liv,lv,v)		build k-d tree
lowesr(y,iv,liv,lv,v)				rebuild with new data values
						(does not change y)
lowese(iv,liv,lv,v,m,z,  s)			evaluate smooth at z
lowesl(iv,liv,lv,v,m,z,  L)			explicit hat matrix,
						which maps from y to z
lofort(iunit,iv,liv,lv,v)			save k-d tree as Fortran
losave(iunit,iv,liv,lv,v)			save k-d tree in file
lohead(iunit,d,vc,nc,nv)			read d,vc,nc,nv from file
	liv = 50+(vc+3)*nc			determine space
	lv = 50+(2*d+1)*nv+nc				requirements
loread(iunit,d,vc,nc,nv,iv,liv,lv,v)		finish reading k-d tree,
							ready for lowese
lowesa(trL,n,d,tau,nsing,  del1,del2)		approximate delta
lowesc(n,L,LL,  trL,del1,del2)			exact delta
lowesp(n,y,yhat,w,rw,  pi,ytilde)		pseudo-values
lowesw(res,n,  rw,pi)				robustness weights

=== arguments ===
d	number of independent variables [integer]  (called "p" elsewhere)
del1,del2	delta1, delta2
diagL	diagonal of hat matrix, only set if infl=.true.    (n)
f	fraction of points to use in local smooth  (called "alpha" elsewhere)
fc	don't refine cells with less than fc*n points;   ordinarily=.05
hat	is hat matrix desired?  [integer]
	0 = none
	1 = diagonal only
	2 = full matrix
infl	is diagonal of hat matrix desired?	[logical]
iunit	Fortran unit number for i/o
iv	workspace  (liv)
L	hat matrix (m,n)   [real]
	in lowesf, only computed if hat nonzero;  if hat=1 only size (n)
LL	workspace (n,n)
liv	50+(2^d+4)*nvmax+2*n
	if setLf, add nf*nvmax
lv	50+(3*d+3)*nvmax+n+(tau0+2)*nf
	if setLf, add (d+1)*nf*nvmax
m	number of points to smooth at;   ordinarily=n
n	number of observations
nf	min(n,floor(n*f))
nsing	if 0, print warning in lowesa when trL<tau;  typically nsing=iv(30)
nvmax	limit on number of vertices for kd-tree; e.g. max(200,n)
pi	workspace (n)  [integer]
res	residual  yhat-y  (n)
rw	robustness weights  (n)
s	smoothed values at z  (m)
setLf	in lowesb, save matrix factorizations  [logical]
	(needed for lowesr and lowesl)
tau	dimension of local model = iv(DIM);
	=d+1 for linear, (d+2)(d+1)/2 for quadratic
		reduced if dropping squares
tau0	=d+1 for linear, (d+2)(d+1)/2 for quadratic
tdeg    polynomials to fit;  0=constants, 1=linear, 2=quadratics
trL	trace L = sum diagL
v	workspace  (lv)
w	weights  (n)    local regression: min sum wi * (f(xi)-yi)^2
x	sample locations  (n,d)
y	observations  (n)
yhat	smoothed y  (n)
ytilde	pseudo y  (n)
z	locations where smooth is desired  (m,d)

If using the double precision version, [real] above should be understood
as Fortran "double precision".

The first argument to lowesd is a version number, updated when calling
sequences change.

If you peek inside the fortran, you will quickly notice that it
was machine generated;  the typeset original (in the language "pine")
is much easier to read.

=== iv indices ===
1	INFO	return code (not currently used)
2	D	number of independent variables
3	N	number of observations
4	VC	2^d  (number of vertices of a cell)
5	NC	number of k-d cells
6	NV	number of k-d vertices
7	A1	starting index in iv of a
8	C1	starting index in iv of c
9	HI1	starting index in iv of hi
10	LO1	starting index in iv of lo
11	V1	starting index in v of vertices
12	XI1	starting index in v of cut values
13	VV1	starting index in v of vertex values
14	NVMAX	maximum allowed value of nv
15	WORK1	starting index in v of workspace
16	WORK2	starting index in v of workspace
17	NCMAX	maximum allowed value of nc
18	WORK3	starting index in v of workspace
19	NF	floor(n*f) (number of points used as neighborhood)
20	KERNEL	1=tricube, 2=unif
21	KIND	1=k-d,cubic blend, (not implemented:2=quadtree,3=triangulation)
22	PI1	starting index in iv of tree permutation
23	VH	starting index in iv of vhit
24	VV2	starting index in v of work vval used in trL computation
25	LQ	starting index in iv of Lq
26	WORK4	starting index in v of workspace
27	PSI1	starting index in iv of workspace permutation
28	SEQ	sequence number, to check if routines called out of order
		takes on values:
		171	after lowesd
		172	after lowesf
		173	after lowesb
29	DIM	dimension of local regression
		1		constant
		d+1		linear   (default)
		(d+2)(d+1)/2	quadratic
		Modified by ehg127 if cdeg<tdeg.
30	SING	number of times singular tolerance was met in l2fit, l2tr
31	PRINT	verbose output?
32	DEG	total degree (of polynomial for local model)
33	NDIST	dd = variables 1:dd enter into distance calculation
34	LF	starting index in v of Lf
35..40		reserved for future use
41..49	CDEG	componentwise degree
iv(A1)	a	coordinate of cut; 0 for leaf  (nc)
iv(C1)	c	pointers to corners (index into vertex array v)  (vc,nc)
iv(HI1)	hi	right subcell  (nc)
iv(LO1)	lo	left subcell  (nc)
		Leaf cell j encloses points x(pi(i),), lo(j)<=i<=hi(j).
		Also, iv(C1),...,iv(PI1-1) is used as workspace (t) by l2fit
------------------------eval only needs workspace up to here
iv(PI1)	pi	permutation of 1:n for listing points in cells
iv(VH)	vhit	cell whose subdivision creates vertex (nv)
		0 if vertex is corner of original bounding box.
iv(LQ)	Lq	active point indices for block of Lf    (nvmax*nf)
iv(PSI1) psi	workspace permutation of 1:n for sorting distances

=== v indices ===
1	F	fraction of n to be used as neighborhood.   See also iv(19).
2	FCELL	no refinement if #points <= fcell * n
		default .05
3	FDIAM	no refinement if diameter is fdiam * overall bounding box
		default 0;    Warning: reset to 0 by ehg142 when nsteps>0.
4	RCOND	reciprocal condition number
... 49		reserved for future use
iv(V1)	v	vertices  (nv,d)
iv(VV1)	vval	vertex values  (0:d,nv)
iv(XI1)	xi	cut values  (nc)
------------------------eval only needs workspace up to here
iv(WORK1)	workspace  (n)  l2fit:dist
iv(WORK2)	workspace  (nf) l2fit:eta
iv(WORK3)	workspace  (dim,nf)   l2fit:X
iv(VV2)	vval2	workspace  ((d+1)*nv)  pseudo-vval for trL computation
iv(LF) 	Lf	hat matrix (data to vertex)   ((d+1)*nvmax*nf)
iv(WORK4)	workspace  (nf) l2fit:w

Internal routine names have been hidden as follows:
ehg106  select q-th smallest by partial sorting
ehg124  rbuild
ehg125  cpvert
ehg126  bbox
ehg127  l2fit,l2tr computational kernel
ehg128  eval
ehg129  spread
ehg131  lowesb after workspace expansion
ehg133  lowese after workspace expansion
ehg134  abort by calling S Recover function
ehg136  l2fit with hat matrix L
ehg137  vleaf
ehg138  descend
ehg139  l2tr
ehg140(w,i,j)   w(i)=j   used when w is declared real, but should store an int
ehg141  delta1,2 from trL
ehg142  robust iteration
ehg144  now called lowesc
ehg152  like ehg142, but for lowesf
ehg167  kernel for losave
ehg168  kernel for loread
ehg169  compute derived k-d tree information
ehg170  generate Fortran
ehg176,ehg177,ehg178,ehg179,ehg180,ehg181  loeval for delta
ehg182  ehgdie(i)
ehg183	warning(message,i,n,inc)
ehg184	warning(message,x,n,inc)
ehg190  now called lowesa, with slight change in calling sequence
ehg191  lowesl after workspace expansion
ehg192  lowesr after workspace expansion
ehg196(tau,d,f,trl)	trL approximation
ehg197    for deg=1,2
m9rwt	now called lowesw
pseudo	now called lowesp
